src/com/linkedin/parseq/BaseTask.java

```diff
@@ -311,12 +320,71 @@ private void done(final T value, final TaskLogger taskLogger) {
 
   private void fail(final Throwable error, final TaskLogger taskLogger) {
     if (transitionDone()) {
+      appendTaskStackTrace(error);
       traceFailure(error);
       getSettableDelegate().fail(error);
       taskLogger.logTaskEnd(BaseTask.this, _traceValueProvider);
     }
   }
 
+  // Concatenate stack traces if kept the original stack trace from the task creation
+  private void appendTaskStackTrace(final Throwable error) {
+    if (!ParSeqGlobalConfiguration.getInstance().isCrossThreadStackTracesEnabled() || error == null ||
+        _taskStackTrace == null || _taskStackTrace.length <= 2) {
+      return;
+    }
+
+    StackTraceElement[] errorStackTrace = error.getStackTrace();
+    if (errorStackTrace.length <= 2) {
+      return;
+    }
+
+    // Skip stack frames up to the BaseTask (useless java.util.concurrent stuff)
+    int skipErrorFrames = 1;
+    while (skipErrorFrames < errorStackTrace.length) {
+      int index = errorStackTrace.length - 1 - skipErrorFrames;
+      if (!errorStackTrace[index].getClassName().equals(CANONICAL_NAME) &&
+          errorStackTrace[index + 1].getClassName().equals(CANONICAL_NAME)) {
+        break;
+      }
+      skipErrorFrames++;
+    }
+
+    // Safeguard against accidentally removing entire stack trace
+    if (skipErrorFrames == errorStackTrace.length) {
+      skipErrorFrames = 0;
+    }
+
+    // Skip stack frames up to the BaseTask (useless Thread.getStackTrace stuff)
+    int skipTaskFrames = 1;
```

Comment:
There's not much actual copy-paste - these two loops go into opposite directions and compute index differently. If I try to combine both logics in a single function it will look a lot more confusing and it will still contain if/else to support bi-directional scan. I would recommend against it.